### PR TITLE
[Modules] Multi additions to staticSites 

### PR DESCRIPTION
**Pull request by MrMCake**
# Description

> Based on PR by @itpropro #1655 

Added the following to staticSites

- `config` (app config and function app config)
- `customDomains` with different validations depending on apex or not
- `linkedBackends` with the 2022-03-01 API to add custom function apps. The `userDefinedFunctions` endpoint should not be used and is currently broken.

Changes were verified, pester and deployment tested with the respective PowerShell functions. Readme and parameter files have been updated.

| Pipeline |
| - |
| [Web: StaticSites](https://github.com/Azure/ResourceModules/actions/workflows/ms.web.staticsites.yml) |

# Type of Change

Please delete options that are not relevant.

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Update to documentation

### CODE REVIEW

File: modules/Microsoft.Web/staticSites/.test/min.parameters.json [hunk 1]
```diff
@@ -1,9 +1,9 @@
 {
-    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
-    "contentVersion": "1.0.0.0",
-    "parameters": {
-        "name": {
-            "value": "<<namePrefix>>-az-wss-min-001"
-        }
+  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
```
Comment (eriqua) [inline modules/Microsoft.Web/staticSites/.test/min.parameters.json:2]: All other json files have configured tab size of 4 instead of 2. Is this change intended?

File: modules/Microsoft.Web/staticSites/.test/parameters.json [hunk 2]
```diff
@@ -1,50 +1,67 @@
...
-            "value": [
-                {
-                    "subnetResourceId": "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.Network/virtualNetworks/adp-<<namePrefix>>-az-vnet-x-001/subnets/<<namePrefix>>-az-subnet-x-005-privateEndpoints",
-                    "service": "staticSites"
-                }
-            ]
-        }
-    }
-}
+{
```
Comment (eriqua) [inline modules/Microsoft.Web/staticSites/.test/parameters.json:1]: All other json files have configured tab size of 4 instead of 2. Is this change intended?

File: modules/Microsoft.Web/staticSites/config/deploy.bicep [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+@description('Optional. Location for all resources.')
+param location string = resourceGroup().location
+
+resource staticSite 'Microsoft.Web/staticSites@2022-03-01' existing = {
+  name: name
+}
+
+resource config 'Microsoft.Web/staticSites/config@2022-03-01' = {
+  #disable-next-line BCP225
```
Comment (eriqua) [inline modules/Microsoft.Web/staticSites/config/deploy.bicep:25]: Could you elaborate on this line?

File: modules/Microsoft.Web/staticSites/config/deploy.bicep [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  'functionappsettings'
+])
+@description('Required. Type of settings to apply.')
+param kind string
+
+@description('Required. App settings.')
+param properties object
+
+@description('Conditional. The name of the parent Static Web App. Required if the template is used in a standalone deployment.')
+param name string
```
Comment (eriqua) [inline modules/Microsoft.Web/staticSites/config/deploy.bicep:12]: For consistency with other children 
```suggestion
param staticSiteName string
```

File: modules/Microsoft.Web/staticSites/config/deploy.bicep [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+@description('Optional. Enable telemetry via the Customer Usage Attribution ID (GUID).')
+param enableDefaultTelemetry bool = true
+
+@description('Optional. Location for all resources.')
+param location string = resourceGroup().location
+
+resource staticSite 'Microsoft.Web/staticSites@2022-03-01' existing = {
+  name: name
```
Comment (eriqua) [inline modules/Microsoft.Web/staticSites/config/deploy.bicep:21]:
```suggestion
  name: staticSiteName
```

File: modules/Microsoft.Web/staticSites/.test/min.parameters.json (hunk 1 above)
Comment (MrMCake) [reply modules/Microsoft.Web/staticSites/.test/min.parameters.json:2]: Must be from the original PR and is technically irrelevant. Will update it manually, but would suggest to ignore the tab size in the future.

File: modules/Microsoft.Web/staticSites/config/deploy.bicep (hunk 3 above)
Comment (MrMCake) [reply modules/Microsoft.Web/staticSites/config/deploy.bicep:25]: Must be from the original PR. If removed Bicep shows a warning: 
image

I'd suggest to keep it

File: modules/Microsoft.Web/staticSites/.test/min.parameters.json (hunk 1 above)
Comment (MrMCake) [reply modules/Microsoft.Web/staticSites/.test/min.parameters.json]: updated

File: modules/Microsoft.Web/staticSites/.test/parameters.json (hunk 2 above)
Comment (MrMCake) [reply modules/Microsoft.Web/staticSites/.test/parameters.json:1]: updated

File: modules/Microsoft.Web/staticSites/config/readme.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+- [Resource Types](#Resource-Types)
+- [Parameters](#Parameters)
+- [Outputs](#Outputs)
+
+## Resource Types
+
+| Resource Type | API Version |
+| :-- | :-- |
+| `Microsoft.Web/staticSites/config` | [2022-03-01](https://docs.microsoft.com/en-us/azure/templates/Microsoft.Web/staticSites/config) |
```
Comment (eriqua) [inline modules/Microsoft.Web/staticSites/config/readme.md:15]: Same as other PR discussion. We're not linking the right API documentation here nor in other added children. Latest is quite older that the one used here and the implementation looks different as well. In the latest documented API, the name property only accepts functionappsettings as a value and kind is a separated property. In this implementation, name and kind properties seem to be mixed, which I cannot be sure about as I lack proper knowledge on this service. Anyway, I find inaccurate to link a different API documentation than the one used. 

I think we should reopen the discussion about what to do if the azure resource reference documentation is outdated or lacking (we also have modules for resources which are not there at all). Probably removing the link to the docs completely and documenting in the Wiki that if links are not there it means that no official documentation in ARR exists yet for the API version we are using.

File: modules/Microsoft.Web/staticSites/.test/min.parameters.json (hunk 1 above)
Comment (eriqua) [reply modules/Microsoft.Web/staticSites/.test/min.parameters.json]: Thanks for the update. I'm not sure I agree on this one. I'd still aim for a consistent formatting througout the repo. Instead of ignoring we could probably avoid it with repo format settings.

File: modules/Microsoft.Web/staticSites/config/readme.md (hunk 6 above)
Comment (MrMCake) [reply modules/Microsoft.Web/staticSites/config/readme.md:15]: Agree. In many cases we cannot link to the documentation because it is frankly not up to date - or not existing at all. While I appreciate that we want to point users to documentation where possible, we should not forget that Bicep comes with its own documentation. I assume in this case, the original contributor @itpropro generatd the documentation (e.g. parameter metadata) from this one.

In any case, I'm also not too familiar with the service and would hence opt-out to refactor the code. There is a fair chance I'd 'improve' it to the worse. 

Tricky one.

File: modules/Microsoft.Web/staticSites/.test/min.parameters.json (hunk 1 above)
Comment (MrMCake) [reply modules/Microsoft.Web/staticSites/.test/min.parameters.json]: If we can control it with a .vscode setting, be my guest. My main concern is usually the correct formatting (e.g. that a bracket is not indented incorrectly or so) instead of the number of indents the entire file follows. E.g. the deploymentExamples don't care at all and will all auto-format the content in the readmes based on its own rules. 

So, if we can control this via a settings, great. If not - I'm not sure if it's worth the fuzz and effort reviewing ^^

File: modules/Microsoft.Web/staticSites/config/readme.md (hunk 6 above)
Comment (itpropro) [reply modules/Microsoft.Web/staticSites/config/readme.md:15]: The latest API version is the one mentioned here, I don't really get the question. The only real source to rely on are the api specs that are documented [here](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/web/resource-manager/Microsoft.Web/stable/2022-03-01/StaticSites.json). They clearly state that the latest stable version is 2022-03-01, which is also the one that fixes a lot of bugs with the older provider like not being able to link a user provided function app for example. 
I think we should always rely on the API specs as they are the source for every SDK, the Azure Portal, the REST APIs and the Docs.

Comment (MrMCake) [reply modules/Microsoft.Web/staticSites/config/readme.md:15]: It is. However, like in the other PR I'm not convinced the Azure Resource Reference is the best reference, especially as it is not updated as often as it should be and should not restrict us. Maybe we can find out why this is the case and contact the team owning it. If not, maybe there is a better reference we can dig up.

Comment (itpropro) [reply modules/Microsoft.Web/staticSites/config/readme.md:15]: Thanks for clarifying @eriqua, I think the API was just released last month and thereby the Docs have not been auto generated. Good idea to ask the responsible team what their plan is on that.

Comment (eriqua) [reply modules/Microsoft.Web/staticSites/config/readme.md:15]: Opened a new issue #1735 to keep track of this discussion

File: modules/Microsoft.Web/staticSites/.test/min.parameters.json (hunk 1 above)
Comment (eriqua) [reply modules/Microsoft.Web/staticSites/.test/min.parameters.json]: PR #1759 should do

File: modules/Microsoft.Web/staticSites/config/readme.md (hunk 6 above)
Comment (eriqua) [reply modules/Microsoft.Web/staticSites/config/readme.md:15]: Version 2022-03-01 is now available in the Azure Resource Reference https://docs.microsoft.com/en-us/azure/templates/microsoft.web/2022-03-01/staticsites?pivots=deployment-language-bicep which would unblock this PR. On the bigger topic we still need to agree on how to proceed when misalignments would raise.

File: modules/Microsoft.Web/staticSites/config/deploy.bicep (hunk 3 above)
Comment (itpropro) [reply modules/Microsoft.Web/staticSites/config/deploy.bicep:25]: This is to circumvent a current limitation of the bicep type inference system.

File: modules/Microsoft.Web/staticSites/.test/min.parameters.json (hunk 1 above)
Comment (itpropro) [reply modules/Microsoft.Web/staticSites/.test/min.parameters.json]: I think we should introduce a linting step as gate to PRs that ensure correct indentation, then the original indentations from the PRs shouldn’t be relevant.

File: modules/Microsoft.Web/staticSites/config/deploy.bicep (hunk 3 above)
Comment (MrMCake) [reply modules/Microsoft.Web/staticSites/config/deploy.bicep:25]: Correct. In a separate PR I'm also adding some context to the Wiki.

File: modules/Microsoft.Web/staticSites/config/readme.md (hunk 6 above)
Comment (MrMCake) [reply modules/Microsoft.Web/staticSites/config/readme.md:15]: I received some feedback and the latest API version should be available now. Once I finished my current efforts I'll wrap this PR up

File: modules/Microsoft.Web/staticSites/.test/min.parameters.json (hunk 1 above)
Comment (MrMCake) [reply modules/Microsoft.Web/staticSites/.test/min.parameters.json]: It appears the parameter file is aligned now. @eriqua, please resolve

File: modules/Microsoft.Web/staticSites/config/deploy.bicep (hunk 3 above)
Comment (MrMCake) [reply modules/Microsoft.Web/staticSites/config/deploy.bicep]: Can this be resolved?

File: modules/Microsoft.Web/staticSites/config/readme.md (hunk 6 above)
Comment (MrMCake) [reply modules/Microsoft.Web/staticSites/config/readme.md:16]: I just updated the ReadMes and triggered the pipeline again. Once (successfully) finished I think we can resolve this comment

File: modules/Microsoft.Web/staticSites/config/deploy.bicep (hunk 3 above)
Comment (itpropro) [reply modules/Microsoft.Web/staticSites/config/deploy.bicep]: I think so @eriqua ?

Comment (eriqua) [reply modules/Microsoft.Web/staticSites/config/deploy.bicep]: As done for other places where we are silencing false positives, we should add a comment explaining why we are disabling the warning, i.e., why the rule does not apply, as suggested [here](https://docs.microsoft.com/en-us/azure/azure-resource-manager/bicep/linter#silencing-false-positives)

**Comment by itpropro** [comment]
@eriqua and @MrMCake can we finally close this PR? The pipeline is failing because of the word "id" is not in uppercase in the linked backends readme, maybe someone can do a quick fix here. Besides that, there is nothing left and I would really like to consume the changes from this module 😅

**Comment by MrMCake** [comment]
Hey @itpropro, I'll take care of it once #1898 is in to make it as easy for me as possible resolving conflicts. Should happen in the next couple days.

**Comment by eriqua** [comment]
> Hey @itpropro, I'll take care of it once #1898 is in to make it as easy for me as possible resolving conflicts. Should happen in the next couple days.

@MrMCake, @itpropro PR #1898 is approved and ready to be merged.

**Comment by MrMCake** [comment]
Merged and retesting

File: modules/Microsoft.Network/privateLinkServices/.test/parameters.json [hunk 7]
```diff
@@ -1,67 +1,66 @@
...
-            ]
-        },
-        "linkedBackend": {
-            "value": {
-                "resourceId": "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.Web/sites/adp-<<namePrefix>>-az-fa-001"
-            }
-        }
-    }
-}
+{
```
Comment (eriqua) [inline modules/Microsoft.Network/privateLinkServices/.test/parameters.json:1]: 👻

File: modules/Microsoft.Web/staticSites/linkedBackends/deploy.bicep [hunk 8]
```diff
@@ -0,0 +1,51 @@
+@description('Required. The resource id of the backend linked to the static site.')
```
Comment (eriqua) [inline modules/Microsoft.Web/staticSites/linkedBackends/deploy.bicep:1]:
```suggestion
@description('Required. The resource ID of the backend linked to the static site.')
```

File: modules/Microsoft.Web/staticSites/linkedBackends/deploy.bicep [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    mode: 'Incremental'
+    template: {
+      '$schema': 'https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#'
+      contentVersion: '1.0.0.0'
+      resources: []
+    }
+  }
+}
+
+@description('The name of the static site.')
```
Comment (eriqua) [inline modules/Microsoft.Web/staticSites/linkedBackends/deploy.bicep:44]:
```suggestion
@description('The name of the static site linked backend.')
```

File: modules/Microsoft.Web/staticSites/linkedBackends/deploy.bicep [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+output name string = linkedBackend.name
+
+@description('The resource ID of the static site.')
```
Comment (eriqua) [inline modules/Microsoft.Web/staticSites/linkedBackends/deploy.bicep:47]:
```suggestion
@description('The resource ID of the static site linked backend.')
```

File: modules/Microsoft.Web/staticSites/linkedBackends/deploy.bicep [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+output resourceId string = linkedBackend.id
+
+@description('The resource group the static site was deployed into.')
```
Comment (eriqua) [inline modules/Microsoft.Web/staticSites/linkedBackends/deploy.bicep:50]:
```suggestion
@description('The resource group the static site linked backend was deployed into.')
```

File: modules/Microsoft.Web/staticSites/customDomains/deploy.bicep [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    mode: 'Incremental'
+    template: {
+      '$schema': 'https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#'
+      contentVersion: '1.0.0.0'
+      resources: []
+    }
+  }
+}
+
+@description('The name of the static site.')
```
Comment (eriqua) [inline modules/Microsoft.Web/staticSites/customDomains/deploy.bicep:40]:
```suggestion
@description('The name of the static site custom domain.')
```

File: modules/Microsoft.Web/staticSites/customDomains/deploy.bicep [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+output name string = customDomain.name
+
+@description('The resource ID of the static site.')
```
Comment (eriqua) [inline modules/Microsoft.Web/staticSites/customDomains/deploy.bicep:43]:
```suggestion
@description('The resource ID of the static site custom domain.')
```

File: modules/Microsoft.Web/staticSites/customDomains/deploy.bicep [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+output resourceId string = customDomain.id
+
+@description('The resource group the static site was deployed into.')
```
Comment (eriqua) [inline modules/Microsoft.Web/staticSites/customDomains/deploy.bicep:46]:
```suggestion
@description('The resource group the static site custom domain was deployed into.')
```

File: modules/Microsoft.Web/staticSites/config/readme.md (hunk 6 above)
Comment (eriqua) [reply modules/Microsoft.Web/staticSites/config/readme.md:16]: @mrmcake, @itpropro brace yourself, I'm afraid we still have the same issue for the config child module, for which latest still points to an outdated version. Apparently the resource reference does not have a unique documentation for the module under the latest API version. It instead provides two, based on the `kind` (although `name` is used instead of `kind`, differently than this implementation): https://learn.microsoft.com/en-us/azure/templates/microsoft.web/2022-03-01/staticsites/config-appsettings?pivots=deployment-language-bicep and https://learn.microsoft.com/en-us/azure/templates/microsoft.web/2022-03-01/staticsites/config-functionappsettings?pivots=deployment-language-bicep. I'm not sure how to proceed from here. In order not to postpone this PR any further, I could suggest to merge it but to keep the discussion open in issue #1735 and maybe call out this specific case with the PG. What's your take on that?

Comment (MrMCake) [reply modules/Microsoft.Web/staticSites/config/readme.md:16]: It's similar with the Sites module I pressume. The problem with the Azure Resource Reference is anyways that it is updated and designed semi-manually - and hence is not always consistent, nor up to date

**Review by eriqua — APPROVED** [review]
🆗
